src/amo/middleware/prefixMiddleware.js

```diff
@@ -16,105 +16,96 @@ export function prefixMiddleware(req, res, next, { _config = config } = {}) {
+    URLPathParts.splice(1, 0, userAgentApp);
     isApplicationFromHeader = true;
-    prependedOrMovedApplication = true;
-  } else if (!hasValidLang) {
-    // If lang wasn't valid or was missing prepend one.
+  } else if (hasValidClientAppInPartOne) {
     log.debug(`Prepending lang to URL: ${lang}`);
     URLPathParts.splice(0, 0, lang);
-    // If we've prepended the lang to the URL we need to re-check our
-    // URL exception and make sure it's valid.
-    hasValidClientAppUrlException = isValidClientAppUrlException(
-      URLPathParts[1],
-      {
-        _config,
-      },
-    );
+  } else if (
+    hasValidClientAppInPartTwo ||
+    hasValidClientAppUrlExceptionInPartTwo
+  ) {
+    log.debug(`Replacing lang in URL: ${URLPathParts[0]} with ${lang}`);
+    URLPathParts.splice(0, 1, lang);
+  } else {
+    log.debug(`Prepending lang and clientApp to URL: ${lang}/${userAgentApp}`);
+    URLPathParts.splice(0, 0, lang, userAgentApp);
+    isApplicationFromHeader = true;
   }
 
-  if (!hasValidClientApp && isValidClientApp(URLPathParts[1], { _config })) {
-    // We skip prepending an app if we'd previously prepended a lang and the
-    // 2nd part of the URL is now a valid app.
-    log.debug('Application in URL is valid following prepending a lang.');
-  } else if (prependedOrMovedApplication) {
-    log.debug(
-      'URL is valid because we added/changed the first part to a clientApp.',
-    );
-  } else if (hasValidLocaleException || hasValidClientAppUrlException) {
-    if (
-      clientAppRoutes.includes(URLPathParts[1]) === false &&
-      (hasValidLang || hasValidLocaleException)
-    ) {
-      log.debug('Exception in URL found; we fallback to addons-server.');
-      // TODO: Remove this once upgraded to react-router 4.
-      res.status(404).end(oneLine`This page does not exist in addons-frontend.
-        Returning 404; this error should trigger upstream (usually
-        addons-server) to return a valid response.`);
-      return next();
-    }
-    log.debug('Exception in URL found; prepending lang to URL.');
-  } else if (!hasValidClientApp) {
-    // If the app supplied is not valid we need to prepend one.
-    log.debug(`Prepending application to URL: ${application}`);
-    URLPathParts.splice(1, 0, application);
-    isApplicationFromHeader = true;
+  const hasValidLocaleException = isValidLocaleUrlException(URLPathParts[2], {
+    _config,
+  });
+  const hasValidClientAppUrlException = isValidClientAppUrlException(
+    URLPathParts[2],
+    {
+      _config,
+    },
+  );
+  if (hasValidClientAppUrlException) {
+    isApplicationFromHeader = false;
+  }
+
+  // clientApp values that are allowed through to the router
+  // TODO: This can be removed when we upgrade to react-router v4.
+  const clientAppRoutes = _config.get('clientAppRoutes');
+
+  if (
+    (hasValidLocaleException || hasValidClientAppUrlException) &&
+    clientAppRoutes.includes(URLPathParts[2]) === false &&
```

Comment:
Nit: this could be `!clientAppRoutes.includes(URLPathParts[2])`

Reply:
yeah, the existing code had `=== false` and I didn't try to understand the unusual choice.  I'll change